tcmur_cmd_handler.c

```diff
@@ -972,102 +974,118 @@ static int xcopy_parse_parameter_list(struct tcmu_device *dev,
 	return ret;
 }
 
-static void handle_xcopy_rw_sync(struct tcmu_device *dev,
-				 struct tcmulib_cmd *cmd,
-				 int ret)
+static int xcopy_read_work_fn(struct tcmu_device *src_dev, struct tcmulib_cmd *cmd);
+static void handle_xcopy_read_cbk(struct tcmu_device *src_dev,
+				  struct tcmulib_cmd *cmd,
+				  int ret);
+
+static void handle_xcopy_write_cbk(struct tcmu_device *dst_dev,
+				  struct tcmulib_cmd *cmd,
+				  int ret)
 {
 	struct xcopy *xcopy = cmd->cmdstate;
+	struct tcmu_device *src_dev = xcopy->src_dev;
+
+	/* write failed - bail out */
+	if (ret != SAM_STAT_GOOD) {
+		tcmu_dev_err(src_dev, "Failed to write to dst device!\n");
+		ret = tcmu_set_sense_data(cmd->sense_buf, MEDIUM_ERROR,
+					  ASC_WRITE_ERROR, NULL);
+		goto finish_err;
+	}
+
+	xcopy->lba_cnt -= xcopy->copy_lbas;
+	if (!xcopy->lba_cnt)
+		goto finish_err;
+
+	xcopy->src_lba += xcopy->copy_lbas;
+	xcopy->dst_lba += xcopy->copy_lbas;
+	xcopy->copy_lbas = min(xcopy->lba_cnt, xcopy->copy_lbas);
+
+	cmd->done = handle_xcopy_read_cbk;
+	ret = async_handle_cmd(xcopy->src_dev, cmd, xcopy_read_work_fn);
+	if (ret != TCMU_ASYNC_HANDLED)
+		goto finish_err;
+
+	return;
 
-	pthread_mutex_lock(&xcopy->lock);
-	xcopy->cbk_finished = true;
-	pthread_cond_signal(&xcopy->cond);
-	pthread_mutex_unlock(&xcopy->lock);
+finish_err:
```

Comment:
This is both the success and failure path right? If so, let's rename the goto so it does not look like a failure path.